eth-rpc/server/servertypes.nim

```diff
@@ -13,15 +14,145 @@ type
+  result = n.getFloat()
+
+proc fromJson(n: JsonNode, argName: string, result: var string) =
+  if n.kind != JString: raise newException(ValueError, "Parameter \"" & argName & "\" expected JString but got " & $n.kind)
+  result = n.getStr()
+
+proc fromJson[T](n: JsonNode, argName: string, result: var seq[T]) =
+  if n.kind != JArray: raise newException(ValueError, "Parameter \"" & argName & "\" expected JArray but got " & $n.kind)
+  result = newSeq[T](n.len)
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[N, T](n: JsonNode, argName: string, result: var array[N, T]) =
+  if n.kind != JArray: raise newException(ValueError, "Parameter \"" & argName & "\" expected JArray but got " & $n.kind)
+  if n.len > result.len: raise newException(ValueError, "Parameter \"" & argName & "\" item count is too big for array")
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[T: object](n: JsonNode, argName: string, result: var T) =
+  if n.kind != JObject: raise newException(ValueError, "Parameter \"" & argName & "\" expected JObject but got " & $n.kind)
+  for k, v in fieldpairs(result):
+    fromJson(n[k], k, v)
+
+proc unpackArg[T](argIdx: int, argName: string, argtype: typedesc[T], args: JsonNode): T =
+  fromJson(args[argIdx], argName, result)
+
+proc setupParams(parameters, paramsIdent: NimNode): NimNode =
+  # Add code to verify input and load parameters into Nim types
+  result = newStmtList()
+  if not parameters.isNil:
+    # initial parameter array length check
+    var expectedLen = parameters.len - 1
+    let expectedStr = "Expected " & $expectedLen & " Json parameter(s) but got "
+    result.add(quote do:
+      if `paramsIdent`.kind != JArray:
+        raise newException(ValueError, "Parameter params expected JArray but got " & $`paramsIdent`.kind)
+      if `paramsIdent`.len != `expectedLen`:
+        raise newException(ValueError, `expectedStr` & $`paramsIdent`.len)
+    )
+    # unpack each parameter and provide assignments
+    for i in 1 ..< parameters.len:
+      let
+        paramName = parameters[i][0]
+        pos = i - 1
+        paramNameStr = $paramName
+        paramType = parameters[i][1]
+      result.add(quote do:
+        var `paramName` = `unpackArg`(`pos`, `paramNameStr`, `paramType`, `paramsIdent`)
+      )
+
+macro multiRemove(s: string, values: varargs[string]): untyped =
+  ## Wrapper for multiReplace
+  var
+    body = newStmtList()
+    multiReplaceCall = newCall(ident"multiReplace", s)
+
+  body.add(newVarStmt(ident"eStr", newStrLitNode("")))
+  let emptyStr = ident"eStr"
+  for item in values:
+    # generate tuples of values with the empty string `eStr`
+    let sItem = $item
+    multiReplaceCall.add(newPar(newStrLitNode(sItem), emptyStr))
+
+  body.add multiReplaceCall
+  result = newBlockStmt(body)
+
+macro on*(server: var RpcServer, path: string, body: untyped): untyped =
```

Comment:
How likely is the pattern
```nim
s.on("path"):
  doSmth
```
vs
```nim
s.on("path") do(someArg: int) -> SomeResult:
  doSmth
```
If `do` is more likely, maybe it is not worth maintaining the other branch, and user's code will be more consistent as well =)
I think it would be beneficial to make `body` typed and expect `nkDo`, or `nkProcDef`, or friends. Then `parameters` would be just `body.params`, so it should be somewhat easier, and more strict as to what the user can feed to your macro.

You can always play with `typed` later, but only if you prohibit the anything-but-do way now. ;)